Code/Tools/FBuild/FBuildCore/Graph/TestNode.cpp

```diff
@@ -52,11 +61,43 @@ bool TestNode::Initialize( NodeGraph & nodeGraph, const BFFIterator & iter, cons
         return false; // InitializePreBuildDependencies will have emitted an error
     }
 
-    // Get node for Executable
-    if ( !function->GetFileNode( nodeGraph, iter, m_TestExecutable, "TestExecutable", m_StaticDependencies ) )
+    // .TestExecutable
+    Dependencies executable;
+    if ( !function->GetFileNode( nodeGraph, iter, m_TestExecutable, "TestExecutable", executable ) )
     {
         return false; // GetFileNode will have emitted an error
     }
+    ASSERT( executable.GetSize() == 1 ); // Should only be possible to be one
+
+    // .TestInput
+    Dependencies testInputFiles;
+    if ( !function->GetFileNodes( nodeGraph, iter, m_TestInput, "TestInput", testInputFiles ) )
+    {
+        return false; // GetFileNodes will have emitted an error
+    }
+
+    // .TestInputPath
+    Dependencies testInputPaths;
+    if ( !function->GetDirectoryListNodeList( nodeGraph,
+                                              iter,
+                                              m_TestInputPath,
+                                              m_TestInputExcludePath,
+                                              m_TestInputExcludedFiles,
+                                              m_TestInputExcludePattern,
+                                              m_TestInputPathRecurse,
+                                              &m_TestInputPattern,
+                                              "TestInputPath",
+                                              testInputPaths ) )
+    {
+        return false; // GetDirectoryListNodeList will have emitted an error
+    }
+    ASSERT( testInputPaths.GetSize() == m_TestInputPath.GetSize() ); // No need to store count since they should be the same
+
+    // Store Static Dependencies
+    m_StaticDependencies.SetCapacity( 1 + testInputFiles.GetSize() + testInputPaths.GetSize() );
+    m_StaticDependencies.Append( executable );
+    m_StaticDependencies.Append( testInputFiles );
+    m_StaticDependencies.Append( testInputPaths );
```

Comment:
We're depending on the directory listings nodes here, but not the actual files found by the directory listing. A DoDynamicDependencies pass is needed to add dynamic dependencies on the results of the DriListNodes. Without this, changes to the contents of those directories will not cause a recompile. CSNode is probably the closest example of what's needed.

Note too that due to a limitation in the dependency engine, deleting a file inside the specified directories won't trigger a rebuild as noted here: https://github.com/fastbuild/fastbuild/issues/107 That's not something you can fix (it will be fixed by a core change to dependency management in the future), but I just wanted to call it out.